syntax = "proto3";
import "google/protobuf/empty.proto";

message CreateUserRequest {
  string mobile = 1;
}

message GetUserByIdRequest {
  uint64 id = 1;
}

message GetUserByMobileRequest {
  string mobile = 1;
}

message UpdateAvatarRequest{
  uint64 id = 1;
  string avatar = 2;
}

message UpdateUsernameRequest{
  uint64 id = 1;
  string username = 2;
}

message UpdatePasswordRequest{
  uint64 id = 1;
  string password = 2;
}

message GetUserPageRequest{
  uint32 page = 1;
  uint32 size = 2;
}

message VerifyUserRequest{
  string mobile = 1;
  string password = 2;
}

message UserInfo{
  uint64 id = 1;
  string mobile = 2;
  string username = 3;
  string avatar = 4;
  bool is_active = 5;
  bool is_staff = 6;
  string last_login = 7;
}

message UserInfoResponse{
  UserInfo user = 1;
}

message UserListResponse{
  repeated UserInfo users = 1;
}

service User{
  rpc CreateUser(CreateUserRequest) returns (UserInfoResponse);
  rpc GetUserById(GetUserByIdRequest) returns (UserInfoResponse);
  rpc GetUserByMobile(GetUserByMobileRequest) returns (UserInfoResponse);
  rpc UpdateAvatar(UpdateAvatarRequest) returns (google.protobuf.Empty);
  rpc UpdateUsername(UpdateUsernameRequest) returns (google.protobuf.Empty);
  rpc UpdatePassword(UpdatePasswordRequest) returns (google.protobuf.Empty);
  rpc GetUserList(GetUserPageRequest) returns (UserListResponse);
  rpc VerifyUser(VerifyUserRequest) returns (UserInfoResponse);
  rpc GetOrCreateUserByMobile(GetUserByMobileRequest) returns (UserInfoResponse);
}